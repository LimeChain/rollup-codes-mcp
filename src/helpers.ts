import fs from 'fs'
import path from 'path';
import matter from 'gray-matter';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

export type CustomChainSpec = {
  opcodes: ChainSpecElementsMap,
  precompiles: ChainSpecElementsMap,
  system_contracts: ChainSpecElementsMap
}

export type ChainSpecElementsMap = {
  [index: string]: ChainSpecElement
}

export type ChainSpecElement = {
  name: string,
  description?: string,
  description2?: string,
  url?: string
}

export type ChainSpecElementStatus = "Unsupported" | "Modified" | "Added"

export type CustomChainSpecWithMeta = {
  chainId: number | null,
  opcodes: ChainSpecElementsMap,
  precompiles: ChainSpecElementsMap,
  system_contracts: ChainSpecElementsMap
}

export const getChainSpec = (network: string, rootDir: string): CustomChainSpecWithMeta => {
  const folder = `${rootDir}/chain-specs/specifications/`;
  let opcodes: ChainSpecElementsMap = {};
  let precompiles: ChainSpecElementsMap = {};
  let system_contracts: ChainSpecElementsMap = {};
  let chainId: number | null = null;
  try {
    const fileContents = fs.readFileSync(`${folder}${network}.json`, 'utf8');
    const chainSpec = JSON.parse(fileContents);
    chainId = chainSpec.chain_id ?? null;
    chainSpec.forks.map((fork: CustomChainSpec) => {
      Object.entries(fork.opcodes || {}).forEach(([opcode, data]) => {
        opcodes[opcode] = data;
      });
      Object.entries(fork.precompiles || {}).forEach(([precompile, data]) => {
        precompiles[precompile] = data;
      });
      Object.entries(fork.system_contracts || {}).forEach(([system_contract, data]) => {
        system_contracts[system_contract] = data;
      });
    });
  } catch (e) {
    console.error('Error in getChainSpec:', e);
  }
  return {
    chainId,
    opcodes,
    precompiles,
    system_contracts,
  };
}

export type MessagingFields = {
  l1ToL2: {
    latency: string | null,
    cost: string | null,
  },
  l2ToL1: {
    latency: string | null,
    cost: string | null,
  }
};

export const getRollupMarkdownFields = (rollupName: string, rootDir: string): {
  blockTime: string | null,
  finality: string | null,
  sequencingFrequency: string | null,
  supportedTransactionTypes: string | null,
  gasLimit: string | null,
  messaging: MessagingFields,
  supportedRpcCalls: Array<{
    method: string,
    params: string,
    rollupBehaviour: string,
    ethereumL1Behaviour: string
  }>
} => {
  const mdxPath = `${rootDir}/src/docs/${rollupName.toLowerCase()}.mdx`;
  let blockTime: string | null = null;
  let finality: string | null = null;
  let sequencingFrequency: string | null = null;
  let supportedTransactionTypes: string | null = null;
  let gasLimit: string | null = null;
  // New fields for messaging
  let messaging: MessagingFields = {
    l1ToL2: {
      latency: null as string | null,
      cost: null as string | null,
    },
    l2ToL1: {
      latency: null as string | null,
      cost: null as string | null,
    },
  };
  let supportedRpcCalls: Array<{
    method: string,
    params: string,
    rollupBehaviour: string,
    ethereumL1Behaviour: string
  }> = [];
  try {
    const mdxContent = fs.readFileSync(mdxPath, 'utf8');
    const blockTimeMatch = mdxContent.match(/<Parameter name="Block\s+Time" value="([^"]+)"/i);
    if (blockTimeMatch) blockTime = blockTimeMatch[1];
    const gasLimitMatch = mdxContent.match(/<Parameter name="Gas Limit" value="([^"]+)"/i);
    if (gasLimitMatch) gasLimit = gasLimitMatch[1];
    const finalityMatch = mdxContent.match(/<Parameter name="Objective Finality"\s+value="([^"]+)"/);
    if (finalityMatch) finality = finalityMatch[1];
    const sequencingFrequencyMatch = mdxContent.match(/<Parameter name="Sequencing Frequency" value="([^"]+)"/);
    if (sequencingFrequencyMatch) sequencingFrequency = sequencingFrequencyMatch[1];
    // Extract supported transaction types as the indented list after the Transaction Types parameter
    const txTypesMatch = mdxContent.match(/<Parameter name="Transaction Types"[^>]*>\s*([\s\S]*?)(?:<|$)/);
    if (txTypesMatch) {
      // Try to extract the markdown list (lines starting with '-') after the parameter
      const listMatch = txTypesMatch[1].match(/(- .*(?:\n|$))+?/g);
      if (listMatch) {
        supportedTransactionTypes = listMatch.map(line => line.trim()).join('\n');
      } else {
        // fallback: just return the raw block
        supportedTransactionTypes = txTypesMatch[1].trim();
      }
    }
    // Extract Messaging MultiRowParameters block
    const messagingMatch = mdxContent.match(/<MultiRowParameters[^>]*title=["']Messaging["'][^>]*data=\{\[([\s\S]*?)\]\}\s*\/>/);
    if (messagingMatch) {
      // Try to parse the JS-like array inside data={[ ... ]}
      const dataBlock = messagingMatch[1];
      // Extract L1 -> L2 and L2 -> L1 blocks
      const l1ToL2Match = dataBlock.match(/title:\s*['"]L1\s*→\s*L2['"],\s*rows:\s*\[([\s\S]*?)\]/);
      const l2ToL1Match = dataBlock.match(/title:\s*['"]L2\s*→\s*L1['"],\s*rows:\s*\[([\s\S]*?)\]/);
      function extractField(rowsBlock: string | undefined, label: string): string | null {
        if (!rowsBlock) return null;
        const match = rowsBlock.match(new RegExp(`label:\\s*['\"]${label}['\"],[^}]*value:\\s*(['\"])(.*?)\\1`));
        return match ? match[2] : null;
      }
      messaging.l1ToL2.latency = extractField(l1ToL2Match ? l1ToL2Match[1] : undefined, 'Latency');
      messaging.l1ToL2.cost = extractField(l1ToL2Match ? l1ToL2Match[1] : undefined, 'Cost');
      messaging.l2ToL1.latency = extractField(l2ToL1Match ? l2ToL1Match[1] : undefined, 'Latency');
      messaging.l2ToL1.cost = extractField(l2ToL1Match ? l2ToL1Match[1] : undefined, 'Cost');
    }
    // Extract supported RPC calls from the RPC-API section
    const rpcSectionMatch = mdxContent.match(/<Section title=["']RPC-API["']>[\s\S]*?<Legend \/>[\s\S]*?(\| Method \|[\s\S]*?\|[\s\S]*?\|)(?:\n\n|<\/Section>|$)/);
    if (rpcSectionMatch) {
      const tableBlock = rpcSectionMatch[1];
      // Split into lines, skip header and separator
      const lines = tableBlock.split('\n').filter(line => line.trim().startsWith('|') && !line.includes(':-----'));
      if (lines.length > 1) {
        // Remove header
        lines.shift();
        for (const line of lines) {
          // | `method` | params | rollup | l1 |
          const cols = line.split('|').map(col => col.trim());
          if (cols.length >= 5) {
            // Remove backticks from method
            const method = cols[1].replace(/`/g, '');
            const params = cols[2];
            const rollupBehaviour = cols[3];
            // Remove <Unsupported />, <Modified />, <Added /> tags from ethereumL1Behaviour
            let ethereumL1Behaviour = cols[4].replace(/<Unsupported\s*\/>|<Modified\s*\/>|<Added\s*\/>/g, '').trim();
            supportedRpcCalls.push({ method, params, rollupBehaviour, ethereumL1Behaviour });
          }
        }
      }
    }
  } catch (e) {
    // File not found or parse error, leave as null
  }
  return { blockTime, finality, sequencingFrequency, supportedTransactionTypes, gasLimit, messaging, supportedRpcCalls };
}

export function getRootDir(importMetaUrl: string): string {
  const __filename = fileURLToPath(importMetaUrl);
  let rootDir = dirname(__filename);
  if (rootDir.includes('/build')) {
    rootDir = rootDir.split('/build')[0];
  } else if (rootDir.includes('/src')) {
    rootDir = rootDir.split('/src')[0];
  }
  return rootDir;
}

/**
 * Lists all rollups by reading .mdx files in the docs directory and extracting title and subtitle.
 * @param rootDir The root directory of the project
 * @returns Array of { name, description } objects for each rollup
 */
export function listRollupsFromDocs(rootDir: string): Array<{ name: string; description: string }> {
  const docsDir = `${rootDir}/src/docs/`;
  let rollups: Array<{ name: string; description: string }> = [];
  try {
    const files = fs.readdirSync(docsDir);
    for (const file of files) {
      if (file.endsWith('.mdx')) {
        const filePath = path.join(docsDir, file);
        const content = fs.readFileSync(filePath, 'utf8');
        const { data } = matter(content);
        if (data.title && data.subtitle) {
          rollups.push({ name: data.title, description: data.subtitle });
        }
      }
    }
  } catch (e) {
    // Let the caller handle errors if needed
    throw e;
  }
  return rollups;
}

export const getRollupCodesName = (rollupName: string): string => {
  const formattedRollupName = rollupName.toLowerCase().replace(/ /g, '-');
  const names = {
    "arbitrum": "arbitrum-one",
    "world": "world-chain",
    "zksync": "zksync-era",
    "zkevm": "polygon-zkevm",
  } as Record<string, string>;
  return names[formattedRollupName] || formattedRollupName;
}