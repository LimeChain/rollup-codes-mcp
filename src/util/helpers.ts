import fs from 'fs';
import tmp from "tmp";
import {simpleGit} from "simple-git";
import { getRollupMarkdownFields } from './markdown.js';
import { CustomChainSpec, ChainSpecElementsMap, CustomChainSpecWithMeta, RollupMap } from '../types.js';

export async function buildRollupCaches() {
  const tmpDir = tmp.dirSync({ unsafeCleanup: true });
  await simpleGit().clone("https://github.com/LimeChain/RollupCodes.git", tmpDir.name, { "--depth": 1, "--branch": "main", "--recursive": null });

  const rollupMap = listRollupsFromChainSpecs(tmpDir.name);
  const supportedRollupsEnum = Object.keys(rollupMap) as [string, ...string[]];

  if (supportedRollupsEnum.length === 0) {
    throw new Error("No rollups found");
  }

  const rollupSpecsCache: Record<string, Record<string, any>> = {};
  Object.entries(rollupMap).forEach(([rollupName, execEnvs]) => {
    for (const execEnv of execEnvs) {
      // If there is only one execEnv, we don't need to specify it
      const optionalExecEnv = (execEnvs.length > 1) ? execEnv : undefined;

      const chainSpec = getChainSpec(rollupName, tmpDir.name, optionalExecEnv);
      const opcodes = Object.entries(chainSpec.opcodes).map(([opcode, data]) => ({ opcode, ...data }));
      const precompiles = Object.entries(chainSpec.precompiles).map(([address, data]) => ({ address, ...data }));
      const systemContracts = Object.entries(chainSpec.system_contracts).map(([address, data]) => ({ address, ...data }));

      const { blockTime, finality, sequencingFrequency, supportedTransactionTypes, gasLimit, messaging, supportedRpcCalls } = getRollupMarkdownFields(rollupName, tmpDir.name);
      
      if (!rollupSpecsCache[rollupName]) {
        rollupSpecsCache[rollupName] = {};
      }

      rollupSpecsCache[rollupName][execEnv]= {
        chainId: chainSpec.chainId,
        description: chainSpec.description,
        opcodes,
        precompiles,
        systemContracts,
        blockTime,
        gasLimit,
        finality,
        sequencingFrequency,
        supportedTransactionTypes,
        messaging,
        supportedRpcCalls,
      };
    }
  });

  fs.rmSync(tmpDir.name, { recursive: true, force: true });
  return { rollupSpecsCache, rollupMap };
}

const getChainSpec = (network: string, rootDir: string, execEnv?: string,): CustomChainSpecWithMeta => {
  const folder = `${rootDir}/chain-specs/specifications/`;
  let opcodes: ChainSpecElementsMap = {};
  let precompiles: ChainSpecElementsMap = {};
  let system_contracts: ChainSpecElementsMap = {};
  let chainId: number | null = null;
  let description: string | null = null;
  try {
    const execEnvSuffix = execEnv ? `_${execEnv}` : "";
    const fileContents = fs.readFileSync(`${folder}${network}${execEnvSuffix}.json`, 'utf8');
    const chainSpec = JSON.parse(fileContents);
    chainId = chainSpec.chain_id ?? null;
    description = chainSpec.description ?? null;
    chainSpec.forks.map((fork: CustomChainSpec) => {
      Object.entries(fork.opcodes || {}).forEach(([opcode, data]) => {
        opcodes[opcode] = data;
      });
      Object.entries(fork.precompiles || {}).forEach(([precompile, data]) => {
        precompiles[precompile] = data;
      });
      Object.entries(fork.system_contracts || {}).forEach(([system_contract, data]) => {
        system_contracts[system_contract] = data;
      });
    });
  } catch (e) {
    console.error('Error in getChainSpec:', e);
  }
  return {
    chainId,
    description,
    opcodes,
    precompiles,
    system_contracts,
  };
}

const listRollupsFromChainSpecs = (rootDir: string): RollupMap => {
  const folder = `${rootDir}/chain-specs/specifications/`;
  const files = fs.readdirSync(folder);

  return files.reduce((acc, file) => {
    const fileName = file.replace('.json', '');
    const [name, execEnv] = fileName.split('_');
    if (!acc[name]) {
      acc[name] = [];
    }
    acc[name].push(execEnv || "evm");
    return acc;
  }, {} as RollupMap);
}