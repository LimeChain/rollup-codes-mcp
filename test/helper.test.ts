import { getRollupCodesName, listRollupsFromDocs } from "../src/helpers.js";
import { describe, it, expect } from "vitest";

describe("listRollupsFromDocs", () => {
  it("should return all rollups with name and description (snapshot)", () => {
    const rollups = listRollupsFromDocs();
    expect(rollups).toMatchSnapshot();
  });
});

describe("getRollupCodesName", () => {
  it("should return the correct rollup name", () => {
    expect(getRollupCodesName("Arbitrum")).toBe("arbitrum-one");
    expect(getRollupCodesName("World")).toBe("world-chain");
    expect(getRollupCodesName("Zksync")).toBe("zksync-era");
    expect(getRollupCodesName("Zkevm")).toBe("polygon-zkevm");
  });
});