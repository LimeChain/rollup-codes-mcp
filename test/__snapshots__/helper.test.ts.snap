// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`listRollupsFromDocs > should return all rollups with name and description (snapshot) 1`] = `
[
  {
    "description": "Abstract is an EVM compatible ZK-Rollup that is built on top of the ZK stack. The protocol features a centralised sequencer operated by the Abstract team",
    "name": "Abstract",
  },
  {
    "description": "Arbitrum is an EVM-compatible Optimistic Rollup. The protocol team centrally operates the sequencer. Fraud proofs are enabled for whitelisted validators. The network also supports WASM-based contracts (MultiVM).",
    "name": "Arbitrum One",
  },
  {
    "description": "Base is an Optimistic Rollup utilizing the OP Stack technology. The protocol features a centralised sequencer operated by the Base team.",
    "name": "Base",
  },
  {
    "description": "Blast is an Optimistic Rollup utilizing the OP Stack technology. It provides native yield for ETH and stablecoins.",
    "name": "Blast",
  },
  {
    "description": "Ink is an Optimistic Rollup utilizing OP Stack technology. The protocol features a centralised sequencer operated by the Ink team.",
    "name": "Ink",
  },
  {
    "description": "Kakarot is a ZK-Rollup composed of an EVM written in Cairo and an RPC layer to allow users to interact with it in an Ethereum format. The EVM is a set of Cairo smart contracts deployed on a StarknetOS chain. At the moment, the rollup uses a trusted and centralised sequencer operated by the core team. Development is currently frozen!",
    "name": "Kakarot",
  },
  {
    "description": "Linea is a ZK-Rollup that targets EVM compatibility. As itâ€™s a product of ConsenSys, it is positioned to potentially benefit from better utilisation of the ecosystem's tools and infrastructure.",
    "name": "Linea",
  },
  {
    "description": "Optimism is an Optimistic Rollup boasting near full equivalence to Ethereum. The protocol features a centralised sequencer operated by the Optimism team. The way the protocol is structured and maintained makes it easy for external teams to reuse the technology and create/deploy their own rollups.",
    "name": "Optimism",
  },
  {
    "description": "Polygon zkEVM is a leading ZK-Rollup that targets bytecode-level compatibility with the Ethereum Virtual Machine. At the moment, the rollup uses a trusted and centralised sequencer operated by the core team. Cryptographic proofs are used to ensure the validity of state transitions within the rollup.",
    "name": "Polygon zkEVM",
  },
  {
    "description": "Scroll is a ZK-Rollup that targets bytecode-level compatibility with the Ethereum Virtual Machine. The protocol features a centralised sequencer operated by the protocol team. The zkEVM circuits are based on Ethereum Foundation's Privacy and Scaling Group research work on ZK-EVMs.",
    "name": "Scroll",
  },
  {
    "description": "Soneium is an Optimistic Rollup utilizing OP Stack technology. The protocol features a centralised sequencer operated by the Soneium team.",
    "name": "Soneium",
  },
  {
    "description": "Taiko aims to be a Based and Type1 equivalent Rollup. It uses tiered multi-proofs system. At present, most Taiko blocks rely on SGX proofs, while the remainder use SP1 and RISC0 proofs. The zkEVM circuits are higher tier proofs not yet enabled in production.",
    "name": "Taiko",
  },
  {
    "description": "World Chain is an Optimistic Rollup utilizing OP Stack technology and made for humans, where the transactions of unique humans have priority and where they also have a free allowance, just for being a World ID verified unique human using proof of personhood. The protocol features a centralised sequencer operated by the World Chain team.",
    "name": "World Chain",
  },
  {
    "description": "Zircuit is an EVM-compatible zero-knowledge rollup with a unique hybrid architecture that combines the OP stack infrastructure with zero-knowledge proofs",
    "name": "Zircuit",
  },
  {
    "description": "ZKsync Era is a ZK-Rollup that targets EVM compatibility in its zk-friendly custom virtual machine. It provides standard JSON RPC API and preserves key EVM features such as smart contract composability while introducing some new concepts such as native account abstraction.",
    "name": "ZKsync Era",
  },
]
`;
